File: ROS/src/roborex/src/library/trajectory_planner_service.py
# Greg Attra
# 03/29/2021

import rospy

from library.probabilistic_roadmap import ProbabilisticRoadmap
from library.grasp_trajectory_planner import GraspTrajectoryPlanner
from library.release_trajectory_planner import ReleaseTrajectoryPlanner
from roborex.srv import TrajectoryPlanner, TrajectoryPlannerRequest, TrajectoryPlannerResponse


class TrajectoryPlannerService():

    def __init__(self):
        self.grasp_planner = GraspTrajectoryPlanner()
        self.release_planner = ReleaseTrajectoryPlanner()
        self.serivice = rospy.Service("trajectory_planner", TrajectoryPlanner, self.callback)

    
    def callback(self, req):
        trajectories = []
        if (req.grasp_trajectory):
            trajectories = self.grasp_planner.plan_trajectory(req.arm_pose, req.target)
        if (req.release_trajectory):
            trajectories = self.release_planner.plan_trajectory(req.arm_pose, req.target)

        res = TrajectoryPlannerResponse()
        res.trajectories = trajectories
        return res

File: RexRobotRetriever/Assets/Scripts/Services/TrajectoryPlannerService.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using RosMessageTypes.Roborex;
using Unity.Robotics.ROSTCPConnector;
using UnityEngine;

public class TrajectoryPlannerService : MonoBehaviour
{
    [SerializeField] private float verticalOffset;
    [SerializeField] private float effOffset;
    [SerializeField] private RobotArmJoints arm;
    [SerializeField] private GameObject target;

    private ROSConnection ros;

    public void Start()
    {
        ros = ROSConnection.instance;
    }

    public void Invoke()
    {
        ArmPose pose = arm.GetArmPose();
        var pos = arm.worldJoint.transform.InverseTransformPoint(target.transform.position);
        Debug.Log(pos);
        TrajectoryPlannerRequest req = new TrajectoryPlannerRequest(
            pose,
            new RosMessageTypes.Geometry.Pose(
                new RosMessageTypes.Geometry.Point(
                    pos.x,
                    pos.y,
                    pos.z),
                 new RosMessageTypes.Geometry.Quaternion(0, 0, 0, 0)),
            true,
            false);
        ros.SendServiceMessage<TrajectoryPlannerResponse>("trajectory_planner", req, Callback);
    }

    public void Callback(TrajectoryPlannerResponse res)
    {
        StartCoroutine(arm.ExecuteTrajectories(res.trajectories));
    }
}
